x/incentives/keeper/distribute_test.go

```diff
@@ -1621,3 +1552,430 @@ func (s *KeeperTestSuite) TestSyncGroupGaugeWeights() {
+// to misconfigure the volume to trigger certain behavior.
+// - shouldSkipFunding - whether the test case should skip funding the module account with the coins needed for the test case to succeed. Used for edge case setup.
+//
+// The structure of the test is as follows:
+// - Set up the environment from the test configuration
+// - Compute the expected distribtuion values based on the test configuration to minimize the number of testcase parameters.
+// - Run system under test
+// - Perform validations
+//
+// Validations include:
+// - checking that groups are skipped in acceptable cases without failing other group distributions
+// - checking that the group gauge is correctly updated - filled epochs increased and distributed coins increased
+// - checking that the internal gauges are correctly updated - coins increased
+func (s *KeeperTestSuite) TestAllocateAcrossGauges() {
+	// consists of 1:1 mapped group and group gauge.
+	type groupConfig struct {
+		group      types.GroupGauge
+		groupGauge types.Gauge
+
+		expectedSkipped bool
+
+		// This is auto-computed and set in the test
+		// set up depending on the test case configuration.
+		expectedTotalDistribution sdk.Coins
+	}
+
+	const (
+		invalidUnderlyingGaugeId = uint64(100)
+	)
+
+	var (
+		// local copies for this test to isolate failures if unexpected
+		// mutations do occur.
+		defaultGroupGauge      = deepCopyGroupGauge(defaultGroupGauge)
+		singleRecordGroupGauge = deepCopyGroupGauge(singleRecordGroupGauge)
+
+		baseTime = time.Date(2021, 1, 1, 0, 0, 0, 0, time.UTC)
+
+		defaultCoins = sdk.NewCoins(sdk.NewCoin("uosmo", osmomath.NewInt(100_000_000)))
+
+		// Volume pre-set configurations.
+		balancerOnlyVolumeConfig  = []osmomath.Int{singleRecordGroupGauge.InternalGaugeInfo.GaugeRecords[0].CumulativeWeight, osmomath.ZeroInt()}
+		balancerAndCLVolumeConfig = []osmomath.Int{defaultGroupGauge.InternalGaugeInfo.GaugeRecords[0].CumulativeWeight, defaultGroupGauge.InternalGaugeInfo.GaugeRecords[1].CumulativeWeight}
+
+		// Remaining coins: 1 x defaultCoins (2x Coins - 1x DistributedCoins)
+		perpetualGauge = types.Gauge{
+			Id:               defaultGroupGaugeId,
+			IsPerpetual:      true,
+			Coins:            defaultCoins.Add(defaultCoins...),
+			StartTime:        baseTime,
+			FilledEpochs:     1,
+			DistributedCoins: defaultCoins,
+		}
+	)
+
+	// 2 changes: flip the isPerpetual flag and set the number of epochs paid over.
+	nonPerpetualGauge := withIsPerpetual(deepCopyGauge(perpetualGauge), false)
+	nonPerpetualGauge.NumEpochsPaidOver = 2
+
+	// Configure to distribute to the invalid underlying gauge that is non-perpetual and finished.
+	groupToInvalidUnderlying := deepCopyGroupGauge(singleRecordGroupGauge)
+	groupToInvalidUnderlying.InternalGaugeInfo.GaugeRecords[0].GaugeId = invalidUnderlyingGaugeId
+
+	////////////////////////////// Test-specific helpers
```

Comment:
Did you consider moving these helpers (and the vars/consts if needed) outside this test to help with readability? Or was the thinking that this config setup should only apply to this test case and therefore should not be abstracted out?

Reply:
Exactly! My line of thinking was that the scope should be restricted to where the helpers are used. I'm in favor of avoiding polluting test-global scope with unnecessary helpers.

This a style guideline I personally prefer. Happy if you'd like to encourage me otherwise